### PR TITLE
More trait error reworking

### PR DESCRIPTION
**Pull request by compiler-errors**
More work on #127492, specifically those sub-bullets under "Move trait error reporting to `error_reporting::traits`". Stacked on top of #127493.

This does introduce new `TypeErrCtxt.*Ext` traits, but those will be deleted soon. Splitting this work into bite-sized pieces is the only way that it's gonna be feasible to both author and review ❤️ 

r? lcnr

### CODE REVIEW

**Comment by compiler-errors** [comment]
This one's ready too :>

File: compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs [hunk 1]
```diff
@@ -127,6 +137,195 @@ pub enum DefIdOrName {
     Name(&'static str),
 }
 
+#[extension(pub trait TypeErrCtxtExt<'a, 'tcx>)]
+impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs:141]: why did you move this out of `fulfillment_errors`?

**Review by lcnr — COMMENTED** [review]
apart from this question r=me

File: compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs:141]: I believe I wanted to separate the public API (`report_fulfillment_errors`) from the impl details (reporting each specific error). Also wanted to make `fulfillment_errors` smaller module, bc it was like 4000 lines long.

Ideally `fulfillment_errors` has all the private impls (reporting specific fulfillment errors) except it's still currently pub because we report a single error (selection errors) manually in coercion rn.

**Comment by compiler-errors** [comment]
question answered, if u want me to fix anything later let's do it in a follow-up, since i want this generally to land before i make the big transition pr

@bors r=lcnr
